

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>getDataFig &mdash; G4 Prediction in human transcriptom 1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="getMainDensities" href="getMainDensities.html" />
    <link rel="prev" title="generateShuffle" href="generateShuffle.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="README.html" class="icon icon-home"> G4 Prediction in human transcriptom
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="countLocation.html">countLocation</a></li>
<li class="toctree-l1"><a class="reference internal" href="countTranscript.html">countTranscript</a></li>
<li class="toctree-l1"><a class="reference internal" href="G4Annotation.html">G4Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="G4AnnotationCirculaire.html">G4AnnotationCirculaire</a></li>
<li class="toctree-l1"><a class="reference internal" href="G4AnnotationPIRNA.html">G4AnnotationPIRNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="G4AnnotationTRNA.html">G4AnnotationTRNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="generateShuffle.html">generateShuffle</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">getDataFig</a></li>
<li class="toctree-l1"><a class="reference internal" href="getMainDensities.html">getMainDensities</a></li>
<li class="toctree-l1"><a class="reference internal" href="readBlast.html">readBlas</a></li>
<li class="toctree-l1"><a class="reference internal" href="readRandom.html">readRandom</a></li>
<li class="toctree-l1"><a class="reference internal" href="recurentFunction.html">recurentFunction</a></li>
<li class="toctree-l1"><a class="reference internal" href="GenerateExonJunctionSequences.html">GenerateExonJunctionSequences</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="README.html">G4 Prediction in human transcriptom</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="README.html">Docs</a> &raquo;</li>
        
      <li>getDataFig</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/getDataFig.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-getDataFig">
<span id="getdatafig"></span><h1>getDataFig<a class="headerlink" href="#module-getDataFig" title="Permalink to this headline">¶</a></h1>
<dl class="docutils">
<dt>Copyright:</dt>
<dd>Copyright Universite of Sherbrooke, departement of biochemistry and
departement     of computation.</dd>
<dt>Date:</dt>
<dd>September 2019</dd>
<dt>Description:</dt>
<dd><dl class="first last docutils">
<dt>With the output file of getMainDensities.py, this script compute all other</dt>
<dd>statistics needed. There is one tsv output by figure.</dd>
</dl>
</dd>
<dt>Command Line:</dt>
<dd>For one chr : python ~/PATH/getDataFig.py</dd>
</dl>
<dl class="function">
<dt id="getDataFig.getFig3Data">
<code class="descclassname">getDataFig.</code><code class="descname">getFig3Data</code><span class="sig-paren">(</span><em>df</em>, <em>path</em><span class="sig-paren">)</span><a class="headerlink" href="#getDataFig.getFig3Data" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the data frame for figure 3.</p>
<p>This function aims to filter the input dataFrame to get information about
pG4r at global level and class level. The final dataFrame contains
information to make lolipop plot for GC content, densities and number
of transcript.
The glaobal statistic are cuomputed for each location and each informations
(GC, densities, percent).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>df</strong> (<em>dataFrame</em>) – contains all informations for all biotype locations.</li>
<li><strong>path</strong> (<em>string</em>) – general path with all files.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">classDf, contains data for a location figure.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dataFrame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="getDataFig.getFigBysubClass">
<code class="descclassname">getDataFig.</code><code class="descname">getFigBysubClass</code><span class="sig-paren">(</span><em>df</em>, <em>path</em>, <em>nameClass</em><span class="sig-paren">)</span><a class="headerlink" href="#getDataFig.getFigBysubClass" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as for figure 3 but for subclass level.</p>
<p>This function aims to filter the input dataFrame to get information about
pG4r at class and subclass level. The final dataFrame contains
information to make lolipop plot for GC content, densities and number
of transcript.
The glaobal statistic are cuomputed for each location and each informations
(GC, densities, percent).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>df</strong> (<em>dataFrame</em>) – contains all informations for all biotype locations.</li>
<li><strong>path</strong> (<em>string</em>) – general path with all files.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">classDf, contains data for a location figure.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dataFrame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="getDataFig.getFigDensityByLocation">
<code class="descclassname">getDataFig.</code><code class="descname">getFigDensityByLocation</code><span class="sig-paren">(</span><em>df</em>, <em>path</em>, <em>nameClass</em><span class="sig-paren">)</span><a class="headerlink" href="#getDataFig.getFigDensityByLocation" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the data frame for figure with pG4r densities by location.</p>
<p>This function aims to filter the input dataFrame to get information about
pG4r at location level. The final dataFrame contains information to make
lolipop plot for GC content, densities and number of location figure for
segmental and point locations
The glaobal statistic are cuomputed for each location and each informations
(GC, densities, percent).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>df</strong> (<em>dataFrame</em>) – contains all informations for all biotype locations.</li>
<li><strong>path</strong> (<em>string</em>) – general path with all files.</li>
<li><strong>nameClass</strong> (<em>string</em>) – name of the class we want to get the data (Coding, LongNC,
Pseudogene).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">classDf, contains data for a location figure.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dataFrame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="getDataFig.removeBiotype">
<code class="descclassname">getDataFig.</code><code class="descname">removeBiotype</code><span class="sig-paren">(</span><em>df</em><span class="sig-paren">)</span><a class="headerlink" href="#getDataFig.removeBiotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes from a data frame, informations of some biotype we don’t want.</p>
<p>A lot of biotypes are removed for mainly 2 reasons : the biotype doesn’t
have enough transcript (less than 50) or the biotype by definition is linked
to a high DNA/RNA variation (like IG genes).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>df</strong> (<em>DataFrame</em>) – origin dataframe with all biotypes.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">updated dataframe with only biotype of interest</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">DataFrame</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="getMainDensities.html" class="btn btn-neutral float-right" title="getMainDensities" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="generateShuffle.html" class="btn btn-neutral float-left" title="generateShuffle" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Sarah Belhamiti, Anaïs Vannutelli

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>